File: csharp/Structural/08-Composite/Composite/CompoundJob.cs
using Composite.Component;

namespace Composite.Composite;

public class CompoundJob : IJobComponent
{
    private readonly List<IJobComponent> _children = new();

    public string Name { get; }

    public decimal Cost => _children.Sum(c => c.Cost);

    public CompoundJob(string name)
    {
        Name = name;
    }

    public void Add(IJobComponent component) => _children.Add(component);
    public void Remove(IJobComponent component) => _children.Remove(component);

    public void Execute()
    {
        Console.WriteLine($"\n🔧 Starting compound job: {Name} (Total: N${Cost:N2})");
        foreach (var child in _children)
            child.Execute();
        Console.WriteLine($"✔️  Compound job complete: {Name}");
    }

    public void Display(int depth = 0)
    {
        Console.WriteLine($"{"".PadLeft(depth * 2)}📁 {Name} [Compound — {_children.Count} tasks, N${Cost:N2}]");
        foreach (var child in _children)
            child.Display(depth + 1);
    }
}

File: typescript/Structural/08-Composite/Composite/CompoundJob.ts
import { IJobComponent } from "../Component/IJobComponent";

export class CompoundJob implements IJobComponent {
  private children: IJobComponent[] = [];

  constructor(readonly name: string) {}

  get cost(): number {
    return this.children.reduce((sum, child) => sum + child.cost, 0);
  }

  add(component: IJobComponent): void {
    this.children.push(component);
  }

  remove(component: IJobComponent): void {
    this.children = this.children.filter((c) => c !== component);
  }

  execute(): void {
    console.log(
      `\n🔧 Starting compound job: ${this.name} (Total: N$${this.cost.toFixed(2)})`,
    );
    for (const child of this.children) child.execute();
    console.log(`✔️  Compound job complete: ${this.name}`);
  }

  display(depth: number = 0): void {
    console.log(
      `${"  ".repeat(depth)}📁 ${this.name} [Compound — ${this.children.length} tasks, N$${this.cost.toFixed(2)}]`,
    );
    for (const child of this.children) child.display(depth + 1);
  }
}
